crates/macro-support/src/parser.rs

```diff
@@ -1153,11 +1158,108 @@ fn function_from_decl(
             generate_typescript: opts.skip_typescript().is_none(),
             generate_jsdoc: opts.skip_jsdoc().is_none(),
             variadic: opts.variadic().is_some(),
+            fn_attrs,
         },
         method_self,
     ))
 }
 
+/// Extracts function attributes
+fn extract_fn_attrs(
+    sig: &mut syn::Signature,
+    attrs: &BindgenAttrs,
+) -> Result<FunctionAttributes, Diagnostic> {
+    let mut args_attrs = vec![];
+    for input in sig.inputs.iter_mut() {
+        if let syn::FnArg::Typed(pat_type) = input {
+            let mut keep = vec![];
+            let mut arg_attrs = FunctionComponentAttributes {
+                ty: None,
+                desc: None,
+                name: None,
+                optional: false,
+            };
+            for attr in pat_type.attrs.iter() {
+                if !attr.path().is_ident("wasm_bindgen") {
+                    keep.push(true);
+                    continue;
+                }
+                keep.push(false);
+                attr.parse_nested_meta(|meta| {
+                    if meta.path.is_ident("js_name") {
+                        if arg_attrs.name.is_some() {
+                            return Err(meta.error("duplicate attribute"));
+                        }
+                        let value = meta.value()?.parse::<syn::LitStr>()?.value();
+                        if is_js_keyword(&value) {
+                            return Err(meta.error("collides with js/ts keywords"));
+                        }
```

Comment:
We should also check that this is a valid JS ident, which I believe we have a function lying around of somewhere.

Reply:
using `BindgenAttrs::find` should already handle that, as it is using the attrgen `js_name` logic

Reply:
swicthed to useing BindgenAttrs::find, so it should already handle that, as it is using the attrgen js_name logic

Reply:
Lets add a UI test for that as well while we are at it.

Reply:
I wasnt able to find an function that validates a js identifier besides `is_js_keyword()`, can you point that for me if it exists?

Reply:
Apparently we don't. But I thought by using `BindgenAttrs::find()` we don't need to do this anymore? Looking at it more closely its because we use `syn::Ident`. While JS is a bit more lax, by allowing `_` at the start and `$` anywhere, its fine for now.

But if you really want to add full support, we should not use `regex`, but [`unicode-ident`](https://crates.io/crates/unicode-ident). See the [specification](https://tc39.es/ecma262/#sec-names-and-keywords) as well.

Reply:
No, for `js_name` we use `String`, not `Ident`.
But I found the relevant function: https://github.com/rustwasm/wasm-bindgen/blob/c9ee617e9a63340bfba0cecd7d53ba5262f25ddc/crates/cli-support/src/js/identifier.rs#L31-L42

Reply:
> No, for `js_name` we use `String`, not `Ident`. But I found the relevant function:
> 
> https://github.com/rustwasm/wasm-bindgen/blob/c9ee617e9a63340bfba0cecd7d53ba5262f25ddc/crates/cli-support/src/js/identifier.rs#L31-L42

ok awesome, as I see this is in cli-support crate, and not in macro-support crate, shall we copy/move this into macro-support crate and use it there for validation? or add a dep of cli-support to macro-support crate and use it there, or should the validation take place in cli-support crate, so I assume if thats the case, it would throw error on using cli bin and not at macro expand time where for example Rustanalyzer can show it

Reply:
I personally like to see it at macro expand time, so maybe this logic can be moved to backend or shared crate, so it can be accessed in multiple places, what do you think?

Reply:
Yes, should definitely happen in the macro. Backend I believe is the right place.

Reply:
Yep, `wasm-bindgen-shared` is only a dependency of the proc-macro (and CLI) but not of `wasm-bindgen` directly. So should be fine!

Reply:
ok so shall I move it to `shared` crate?

Reply:
Yes please! Thanks!

Reply:
moved to `shared` crate